### DISCUSSION TITLE
Seed SQL with variables

### DISCUSSION THREAD

**Discussion by 5hee75** [supabase/supabase#9363] [Questions] [answered]
I don't have a ton of SQL experience (I come more from a FE background), but I am trying to set up my `seed.sql` file to give myself some useful test data and keep running into this very verbose pattern of inserting data into one table and then trying to reference fields in that data when seeding other tables.

Simple (contrived) example:

```sql
INSERT INTO public.accounts (name) VALUES ('Account One'), ('Account Two');

INSERT INTO public.contacts (account_id, first_name, last_name)
VALUES
    ((SELECT id FROM public.accounts WHERE name = 'Account One'), 'John', 'Smith'),
    ((SELECT id FROM public.accounts WHERE name = 'Account One'), 'Bob', 'Jones');
```

I've got a lot of tables with foreign keys, so I'd like to be able to save various column values into variables after they are inserted and then use those values through the seed file.

Example:

```sql
INSERT INTO public.accounts (name) VALUES ('Account One'), ('Account Two');
SET @account_1_id = (SELECT id FROM public.accounts WHERE name = 'Account One');

INSERT INTO public.contacts (account_id, first_name, last_name)
VALUES
    (@account_1_id, 'John', 'Smith'),
    (@account_1_id, 'Bob', 'Jones');
```

Maybe I should just be explicitly defining all of the values that I want to be referencing, but that seems very brittle if I ever want to change the seed data.  I really don't like repeating myself or hard-coding more than is necessary.

I've tried `SET var = (SELECT col FROM ...)` and `SELECT var  = col FROM ...` and `SELECT col INTO var FROM ...` but I keep getting syntax errors around my variables, so I'm wondering if `seed.sql` can even support this type of variable declaration?

**Comment by steve-chavez** [comment, marked as answer]
> I've tried SET var = (SELECT col FROM ...) and SELECT var = col FROM ... and SELECT col INTO var FROM ...

Those are `plpgsql` statements. You'd need a DO block to use them inside a sql file

- https://www.postgresqltutorial.com/postgresql-plpgsql/plpgsql-block-structure/
- https://www.postgresql.org/docs/current/sql-do.html

If you want pure SQL you can try https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-MODIFYING

**Reply by 5hee75** [reply]
I see! That makes more sense then. So is it reasonable to wrap the entire seed.sql file in a DO block then and treat it like plpgsql? Is there a downside to that?

**Reply by steve-chavez** [reply]
Hm, I find plain SQL more maintainable but besides from that I don't see any other downside.
